### ISSUE TITLE
ObjectPattern instead of ObjectExpression inside BinaryExpression

### ISSUE THREAD

**Issue by azz** [eslint/typescript-eslint-parser#331]
**What version of TypeScript are you using?**

`2.4.0`

**What version of `typescript-eslint-parser` are you using?**

`ts-2.4` branch

**What code were you trying to parse?**

```js
(a => ({})) + 1;
```

**What did you expect to happen?**

The `{}` should be an `ObjectExpression`.

**What happened?**

The `{}` is an `ObjectPattern`.

This is only an issue inside `BinaryExpression`s, and interestingly is not a problem with `1 + (a => ({}));`.

See [repro here](https://prettier.github.io/prettier/#%7B%22content%22%3A%22(a%20%3D%3E%20(%7B%7D))%20%2B%201%22%2C%22options%22%3A%7B%22printWidth%22%3A80%2C%22tabWidth%22%3A2%2C%22singleQuote%22%3Afalse%2C%22trailingComma%22%3A%22none%22%2C%22bracketSpacing%22%3Atrue%2C%22jsxBracketSameLine%22%3Afalse%2C%22parser%22%3A%22typescript%22%2C%22semi%22%3Atrue%2C%22useTabs%22%3Afalse%2C%22doc%22%3Afalse%2C%22ast%22%3Atrue%7D%7D).

**Comment by azz** [comment]
Root cause seems to be this line:
https://github.com/eslint/typescript-eslint-parser/blob/fb66f615a313d782cf95853699cb1155613ed9fb/lib/convert.js#L616

**Comment by azz** [comment]
I'd attempt to fix this but I'm don't know what that line was originally added for.

**Comment by soda0289** [comment]
Thanks for the report. I can look into this on the weekend.
